File: go/array_hashing/347_top_k_frequent_elements.go
package main

import "fmt"

func main() {
	// Test cases
	// result: [1, 2]
	fmt.Println(topKFrequent([]int{1, 1, 1, 2, 2, 3}, 2))

	// result: [1]
	fmt.Println(topKFrequent([]int{1}, 1))

	// result: [5, 2, 1]
	fmt.Println(topKFrequent([]int{10, 1, 1, 1, 1, 1, 5, 5, 5, 5, 2, 2, 2, 1, 0}, 3))
}

func topKFrequent(nums []int, k int) []int {
	count := make(map[int]int)
	for i := 0; i < len(nums); i++ {
		count[nums[i]] += 1
	}

	frequency := make([][]int, len(nums)+1)

	for num, count := range count {
		frequency[count] = append(frequency[count], num)
	}

	result := make([]int, 0, k)
	for i := len(frequency) - 1; i >= 0; i-- {
		if k <= 0 {
			return result
		}
		if len(frequency[i]) <= 0 {
			continue
		}
		k -= len(frequency[i])
		result = append(result, frequency[i]...)
	}

	return result
}

File: js/array_hashing/347_top_k_frequent_elements.ts
function topKFrequent(nums: number[], k: number): number[] {
    const count: Map<number, number> = new Map()
    for (const num of nums) {
        count.set(num, (count.get(num) || 0) + 1)
    }

    const frequency: number[][] = []
    for (const [n, c] of count) {
        frequency[c] = [...frequency[c] || [], n]
    }

    const result: number[] = []
    for (const f of frequency.reverse()) {
        if (k <= 0) {
            return result
        }
        if (f && f.length) {
            k = k - f.length
            result.push(...f)
        }
    }
    return result
}

(() => {
    // Test cases

    // result: [1, 2]
    console.log(topKFrequent([1, 1, 1, 2, 2, 3], 2))

    // result: [1]
    console.log(topKFrequent([1], 1))

    // result: [5, 2, 1]
    console.log(topKFrequent([10, 1, 1, 1, 1, 1, 5, 5, 5, 5, 2, 2, 2, 1, 0], 3))
})()
